### PR TITLE
footer statistics

### LINKED ISSUE #196: add statistics
статистику добавить в футер. типа 
100 сниппетов
66 тегов
12 авторов

### PR DESCRIPTION
**Pull request by fetis**
Adds statistics information in the footer

Screenshot_2019-06-25 30 Seconds of angular

closes #196

### CODE REVIEW

File: builder/plugins/statistics.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+const exec = require('child_process').exec;
+
+function getSnippetsCount(metadata) {
+  return metadata.snippets.length;
+}
+
+function getAuthorsCount(metadata) {
+  const set = new Set();
```
Comment (NothingEverHappens) [inline builder/plugins/statistics.js]: nit: just do smth like `new Set(metadata.snippets.map(({author})=>({author}))).size;`

File: builder/plugins/statistics.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  metadata.snippets.forEach(snippet => set.add(snippet.author));
+
+  return set.size;
+}
+
+function getTagsCount(metadata) {
```
Comment (NothingEverHappens) [inline builder/plugins/statistics.js]: nit: getTagsCount({tags}) {

**Comment by fetis** [comment]
add commit date to footer

File: builder/plugins/statistics.js (hunk 2 above)
Comment (NothingEverHappens) [reply builder/plugins/statistics.js]: Neither is resolved> DId you push the changes?

**Comment by fetis** [comment]
review notes and added test.
@NothingEverHappens plz re-check

File: builder/__tests__/plugins/statistics.test.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      tags: {
+        tag1: [],
+        tag2: []
+      }
+    }
+  }
+});
+
+test('statistics function', done => {
+  statisticsFn(testFiles, testMetalsmith, () => {
```
Comment (kirjs) [inline builder/__tests__/plugins/statistics.test.js:38]: Can we call statisticsFn consistent with how it's used?

File: builder/__tests__/plugins/statistics.test.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    const { _metadata: { statistics} } = testMetalsmith;
```
Comment (kirjs) [inline builder/__tests__/plugins/statistics.test.js:39]: What happens if the callback never gets called? will it actually timeout?

File: builder/__tests__/plugins/statistics.test.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    expect(statistics.snippets).toBe(3);
+    expect(statistics.authors).toBe(2);
+    expect(statistics.tags).toBe(2);
+    expect(statistics.revision).toBe('hash123');
+    expect(statistics.revisionAbbr).toBe('hsh');
+    expect(statistics.revisionDate).toBe('2019-06-07');
+    expect(statistics.generationDate).toMatch(/\d{4,}-\d\d-\d\d/);
```
Comment (kirjs) [inline builder/__tests__/plugins/statistics.test.js:47]: just do .toMatchSnapshot() ?

File: builder/partials/body.hbs [hunk 6]
```diff
@@ -29,14 +29,19 @@
...
-        }
-    </style>
+    <div class="card">
+        {{#with statistics }}
+            <p>
+                Statistics: {{ snippets }} snippets, {{ authors }} authors, {{ tags }} tags.
+            </p>
+            <p>
+                Generated {{ generationDate }}, based on commit
+                <a href="https://github.com/nycJSorg/30-seconds-of-angular/tree/{{ revision }}" target="_blank">{{ revisionAbbr }}</a>
```
Comment (kirjs) [inline builder/partials/body.hbs:39]: Can we have it in site metadata? We're going to move the repo soon

File: builder/public/css/style.css [hunk 7]
```diff
@@ -121,5 +121,26 @@ p > code {
...
-  height: 70px;
+  margin-top: 1em;
+  min-height: 80px;
+  color: #999;
 }
+footer p {
+  margin: 0;
+}
+
+footer p + p{
```
Comment (kirjs) [inline builder/public/css/style.css]: format this file

**Review by kirjs — CHANGES_REQUESTED** [review]

File: builder/partials/body.hbs (hunk 6 above)
Comment (fetis) [reply builder/partials/body.hbs:39]: I don't think this information belongs to metalsmith meta. It's much easier to change link in template.

File: builder/__tests__/plugins/statistics.test.js (hunk 5 above)
Comment (fetis) [reply builder/__tests__/plugins/statistics.test.js:47]: > just do .toMatchSnapshot() ?

Generation date will be different every time, so snapshot will not match

File: builder/__tests__/plugins/statistics.test.js (hunk 3 above)
Comment (fetis) [reply builder/__tests__/plugins/statistics.test.js:38]: > Can we call statisticsFn consistent with how it's used?

what do you mean? that is how it's used by metalsmith

File: builder/__tests__/plugins/statistics.test.js (hunk 4 above)
Comment (fetis) [reply builder/__tests__/plugins/statistics.test.js:39]: ok. you're asking for error case test

File: builder/__tests__/plugins/statistics.test.js (hunk 5 above)
Comment (kirjs) [reply builder/__tests__/plugins/statistics.test.js:47]: would something like this work?
```
.toMatchSnapshot({
  generationDate: expect.match(/../),
});
```

File: builder/__tests__/plugins/statistics.test.js (hunk 3 above)
Comment (kirjs) [reply builder/__tests__/plugins/statistics.test.js:38]: you don't call it `statisticsFn` when you use it

File: builder/partials/body.hbs (hunk 6 above)
Comment (kirjs) [reply builder/partials/body.hbs:39]: `https://github.com/nycJSorg/30-seconds-of-angular ` this part certainly should not be it 10 different templates

**Review by kirjs — APPROVED** [review]

**Pull request merged**